crates/sui-graphql-rpc/src/types/balance.rs

```diff
@@ -110,37 +117,75 @@ impl Balance {
 
         Ok(conn)
     }
-}
 
-impl Target<Cursor> for StoredBalance {
-    type Source = objects::table;
+    fn base_query(address: SuiAddress, coin_type: Option<TypeTag>, lhs: i64, rhs: i64) -> RawQuery {
+        let mut snapshot_objs = query!(r#"SELECT * FROM objects_snapshot"#);
+        snapshot_objs = Balance::filter(snapshot_objs, address, coin_type.clone());
+
+        let mut history_objs = query!(r#"SELECT * FROM objects_history"#);
+        history_objs = Balance::filter(history_objs, address, coin_type.clone());
+        history_objs = filter!(
+            history_objs,
+            format!(r#"checkpoint_sequence_number BETWEEN {} AND {}"#, lhs, rhs)
+        );
+
+        let final_ = consistent_object_read(snapshot_objs, history_objs, lhs, rhs);
+        final_.group_by("candidates.coin_type")
+    }
 
-    fn filter_ge<ST, GB>(cursor: &Cursor, query: Query<ST, GB>) -> Query<ST, GB> {
-        query.filter(objects::dsl::coin_type.ge((**cursor).clone()))
+    fn filter(mut query: RawQuery, address: SuiAddress, coin_type: Option<TypeTag>) -> RawQuery {
+        query = filter!(
+            query,
+            format!(
+                "owner_id = '\\x{}'::bytea AND coin_type IS NOT NULL",
+                hex::encode(address.into_vec())
+            )
+        );
+
+        if let Some(coin_type) = coin_type {
+            query = filter!(
+                query,
+                "coin_type = {}",
+                coin_type.to_canonical_display(/* with_prefix */ true)
+            );
+        };
+
+        query
     }
+}
 
-    fn filter_le<ST, GB>(cursor: &Cursor, query: Query<ST, GB>) -> Query<ST, GB> {
-        query.filter(objects::dsl::coin_type.le((**cursor).clone()))
+impl RawPaginated<Cursor> for StoredBalance {
+    fn filter_ge(cursor: &Cursor, query: RawQuery) -> RawQuery {
+        filter!(query, "candidates.coin_type >= {}", (**cursor).clone())
     }
 
-    fn order<ST, GB>(asc: bool, query: Query<ST, GB>) -> Query<ST, GB> {
-        use objects::dsl;
-        if asc {
-            query.order_by(dsl::coin_type.asc())
-        } else {
-            query.order_by(dsl::coin_type.desc())
+    fn filter_le(cursor: &Cursor, query: RawQuery) -> RawQuery {
+        filter!(query, "candidates.coin_type <= {}", (**cursor).clone())
+    }
+
+    fn order(asc: bool, query: RawQuery) -> RawQuery {
+        match asc {
```

Comment:
This is a really strange way to write `if`.

Reply:
lol i must be losing my mind